ast/flag.go

```diff
@@ -0,0 +1,219 @@
+// Copyright 2015 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ast
+
+const preEvaluable = FlagHasParamMarker | FlagHasFunc | FlagHasVariable
+
+// IsPreEvaluable checks if the expression can be evaluated before execution.
+func IsPreEvaluable(expr ExprNode) bool {
+	return expr.GetFlag()|preEvaluable == preEvaluable
+}
+
+// IsConstant check if the expression is constant.
+// A constant expression is safe to be rewritten to value expression.
+func IsConstant(expr ExprNode) bool {
+	return expr.GetFlag() == FlagConstant
+}
+
+// SetFlag sets flag for expression.
+func SetFlag(n Node) {
+	var setter flagSetter
+	n.Accept(&setter)
+}
+
+type flagSetter struct {
+}
+
+func (f *flagSetter) Enter(in Node) (Node, bool) {
+	return in, false
+}
+
+func (f *flagSetter) Leave(in Node) (Node, bool) {
+	switch x := in.(type) {
+	case *ValueExpr:
+	case *BetweenExpr:
+		x.SetFlag(x.Expr.GetFlag() | x.Left.GetFlag() | x.Right.GetFlag())
+	case *BinaryOperationExpr:
+		x.SetFlag(x.L.GetFlag() | x.R.GetFlag())
+	case *CaseExpr:
+		f.caseExpr(x)
+	case *SubqueryExpr:
+		x.SetFlag(FlagHasSubquery)
+	case *CompareSubqueryExpr:
+		x.SetFlag(x.L.GetFlag() | x.R.GetFlag())
+	case *ColumnNameExpr:
+		x.SetFlag(FlagHasReference)
+	case *DefaultExpr:
```

Comment:
no need to handle Default?
The default value may be dynamic like timestamp type.